Detect and record large collections

bhalevy opened the pull request:
This series adds support for detecting collections that have too many items
and recording them in `system.large_cells`.

A configuration variable was added to db/config: `compaction_collection_items_count_warning_threshold` set by default to 10000.
Collections that have more items than this threshold will be warned about and will be recorded as a large cell in the `system.large_cells` table.  Documentation has been updated respectively.

A new column was added to system.large_cells: `collection_items`.
Similar to the `rows` column in system.large_partition, `collection_items` holds the number of items in a collection when the large cell is a collection, or 0 if it isn't.  Note that the collection may be recorded in system.large_cells either due to its size, like any other cell, and/or due to the number of items in it, if it cross the said threshold.

Note that #11449 called for a new system.large_collections table, but extending system.large_cells follows the logic of system.large_partitions is a smaller change overall, hence it was preferred.

Since the system keyspace schema is hard coded, the schema version of system.large_cells was bumped, and since the change is not backward compatible, we added a cluster feature - `LARGE_COLLECTION_DETECTION` - to enable using it.
The large_data_handler large cell detection record function will populate the new column only when the new cluster feature is enabled.

In addition, unit tests were added in sstable_3_x_test for testing large cells detection by cell size, and large_collection detection by the number of items.

Closes #11449

bhalevy commented:
Converted to Draft until #11654 is merged and this PR will be rebased on top of it (currently it contains the dependent series)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2461/

bhalevy commented:
rebased

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2473/

bhalevy commented:
In v3 ([165e5b6](https://github.com/scylladb/scylladb/commit/165e5b6c2b5c4dd62a400266155b862c06b0b18a)):
- Added sstable large data stat entry for items_in_collection
  - holds threshold and the max number of collection items in the sstable
  - used for deleting the sstable from system.large_cells when the sstable is deleted
  - updated also docs and tools/scylla-sstable about it

bhalevy commented:
Converted to draft for validating with dtest

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2478/

bhalevy commented:
In v4 ([a564366](https://github.com/scylladb/scylladb/commit/a5643669cb635a1b892e0f4cea8b607154b97949)):
- large_data_handler: keep `_feat.large_collection_detection.when_enabled` listener as member
- mx/writer: maintain both cell_size vs. items_in_collection large data stats above_threshold counter
- added debug log before deletion of large data entries

bhalevy commented:
passed wide_rows_test dtests plus an additional test for large collections (depends on this PR)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2479/

avikivity commented on db/config.cc:
```diff
@@ -382,6 +382,8 @@ db::config::config(std::shared_ptr<db::extensions> exts)
     , compaction_rows_count_warning_threshold(this, "compaction_rows_count_warning_threshold", value_status::Used, 100000,
         "Log a warning when writing a number of rows larger than this value")
+    , compaction_collection_items_count_warning_threshold(this, "compaction_collection_items_count_warning_threshold", value_status::Used, 10000,
+        "Log a warning when writing a collection containing more items than this value")
```
Let's make this LiveUpdate so a user can change the threshold without restarting the server.

avikivity replied:
We can do this as a follow-up, since related config items also need changing.

bhalevy replied:
https://github.com/scylladb/scylladb/issues/11685

avikivity commented on docs/dev/sstable-scylla-format.md:
```diff
@@ -129,6 +129,7 @@ guaranteed to be disjoint (non-overlapping) in their partition keys.
         row_size = be32(2)          // row size, in bytes
         cell_size = be32(3)         // cell size, in bytes
         rows_in_partition = be32(4) // number of rows in a partition
+        items_in_collection = be32(5) // number of items in a collection
```
Need to update `large_data_type`

bhalevy replied:
it is updated. In the last hunk of this patch.

avikivity replied:
It also appears a few rows above.

bhalevy replied:
ok, got it now

bhalevy replied:
fixed in v5

avikivity commented on tools/scylla-sstable.cc:
```diff
@@ -1344,6 +1344,7 @@ const char* to_string(sstables::large_data_type t) {
         case sstables::large_data_type::row_size: return "row_size";
         case sstables::large_data_type::cell_size: return "cell_size";
         case sstables::large_data_type::rows_in_partition: return "rows_in_partition";
+        case sstables::large_data_type::items_in_collection: return "items_in_collection";
```
This should be part of the previous commit.

bhalevy replied:
done in v5

avikivity commented on sstables/mx/writer.cc:
```diff
@@ -991,13 +991,23 @@ void writer::maybe_record_large_rows(const sstables::sstable& sst, const sstable
+        if (collection_items > collection_items_entry.threshold) {
+            collection_items_entry.above_threshold++;
+        }
     };
```
Note: this looks very inefficient, looking up something in a hash only to not do anything with it.

bhalevy replied:
We conditionally update the entries' `max_value` after the lookup. That hasn't changed.
Maybe it should have been a disk array to save the map lookup, but then it would be harder to extend as a disk format.

avikivity replied:
Sure, it was inefficient before that patch too.

We should use separate member variables, and convert to disk_hash only when writing at the end.

bhalevy replied:
ok, will do as follow up

bhalevy replied:
https://github.com/scylladb/scylladb/issues/11686

avikivity commented on db/large_data_handler.cc:
```diff
@@ -95,7 +95,7 @@ future<> large_data_handler::maybe_delete_large_data_entries(sstables::shared_ss
     }
     future<> large_cells = make_ready_future<>();
-    if (above_threshold(ldt::cell_size)) {
+    if (above_threshold(ldt::cell_size) || above_threshold(ldt::items_in_collection)) {
```
This can be merged into the patch that adds the writing. This is really an artificial split.

bhalevy replied:
done in v5

bhalevy commented:
In v5 ([89c62e0](https://github.com/scylladb/scylladb/commit/89c62e027defbdca80e5c41207207d874acee604))
- docs/dev/sstable-scylla-format.md: added `items_in_collection` to `large_data_type` as another option
- "tools: scylla-sstable: print sstables::large_data_type::items_in_collection": squashed with previous patch
- "db/large_data_handler: maybe_delete_large_data_entries: consider sstable items_in_collection stat": squashed into "db/large_data_handler: cql_table_large_data_handler: record large_collections"
- retested

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2486/

denesb commented on conf/scylla.yaml:
```diff
@@ -410,6 +410,9 @@ commitlog_total_space_in_mb: -1
 # compaction_rows_count_warning_threshold: 100000
 
+# Log a warning when writing a collection containing more items than this value
+# compaction_collection_items_count_warning_threshold: 10000
```
I think using `element` instead of `item` would be better.

denesb reviewed (commented):
I think this PR needs to replace `item` with `element` but otherwise LGTM.

bhalevy commented:
In v6 ([8248f08](https://github.com/scylladb/scylladb/commit/8248f08438303e30790f38510b85c7e8b55bec9c)):
- s/items/elements/g

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2510/

bhalevy commented:
In v7 ([021b8dc](https://github.com/scylladb/scylladb/commit/021b8dcdb84a201ce782daae444f03f2d6ab7f6b)):
- fixed over-eager items -> elements changes to unrelated comments
- utils/updateable_value: add simple_value_updater
- db: large_data_handler: dynamically update config thresholds

bhalevy commented:
(sorry @denesb, a couple more patches while we're at it)

In v8 (87f0c9082d62256136a71ad9a7263607b467a187):
- sstables: mx/writer: keep large data stats entry as members
- sstables: mx/writer: optimize large data stats members order

denesb commented on sstables/mx/writer.cc:
```diff
@@ -1470,7 +1467,13 @@ void writer::consume_end_of_stream() {
+    lds.map.emplace(large_data_type::cell_size, std::move(_cell_size_entry));
+    lds.map.emplace(large_data_type::rows_in_partition, std::move(_rows_in_partition_entry));
+    lds.map.emplace(large_data_type::elements_in_collection, std::move(_elements_in_collection_entry));
+    std::optional<scylla_metadata::large_data_stats> ld_stats(std::move(lds));
```
nit: why not make it an optional in the first place?

bhalevy replied:
hmm, yes, this can be done.
my first version called the automatic variable `_large_data_stats` to mimic the previous member so to reduce the code churn, but then I decided to rename it, voiding the churn benefits  so you're right, we can just populate ld_stats after it's created empty.

denesb commented on sstables/mx/writer.cc:
```diff
@@ -594,9 +594,9 @@ class writer : public sstable_writer::writer_impl {
+    large_data_stats_entry _rows_in_partition_entry;
     large_data_stats_entry _row_size_entry;
     large_data_stats_entry _cell_size_entry;
-    large_data_stats_entry _rows_in_partition_entry;
```
Not gonna object but seems like premature optimization to me.

bhalevy replied:
we do access it frequently when writing the sstable so I think it's worth it (although I have no benchmark to prove it)

denesb replied:
Ok, as this doesn't make the code any harder to read/modify I won't ask for one.

denesb reviewed (approved):
LGTM, one nit.

bhalevy commented:
In v9 ([0453a36](https://github.com/scylladb/scylladb/commit/0453a3650c77332436df7099daa158b395d1fc73)):
- prepare `std::optional<scylla_metadata::large_data_stats>` in one go in `writer::consume_end_of_stream`

avikivity commented on utils/updateable_value.hh:
```diff
@@ -210,4 +211,23 @@ updateable_value<T>::observe(std::function<void (const T&)> callback) const {
+    utils::observer<UpdateableValueType> _observer;
+
+public:
+    simple_value_updater(ValueType& value, utils::updateable_value<UpdateableValueType> updateable_value, ValueType factor = 1)
```
Why is the important name simple_value_updater used for multiplying?

I can see transforming_value_updater and unit_converting_value_updater.

bhalevy replied:
Changed to `transforming_value_updater` in v10 (using a std::function for the optional transform function)

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2554/

annastuchlik commented:
@tzach Should this PR be backported to branch 5.1?

bhalevy commented:
> @tzach Should this PR be backported to branch 5.1?

In principle, we don't backport features, just fixes.
However this case might be exceptional since those large collection may cause performance issues so detecting can help prevent that.  @avikivity what do you think?
